src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingWatcher.cs

```diff
@@ -129,93 +108,44 @@ private FileChangeList ComputeChangesAndUpdateState()
-            {
-                if (start[i] != possibleEnding[i]) return false;
-            }
-            return true;
-        }
-
-        private unsafe bool IsWatched(char* filename)
+        internal bool IsWatched(ref FileSystemEntry entry)
         {
+            if (entry.IsDirectory) return false;
             if (_extensionsToWatch == null) return true;
-            var length = GetLength(filename);
-            foreach(var extension in _extensionsToWatch)
+            foreach (var extension in _extensionsToWatch)
             {
-                if(EndsWith(filename, length, extension))
-                {
+                var ignoreCase = !RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
+                if (FileSystemName.MatchesSimpleExpression(extension, entry.FileName, ignoreCase: ignoreCase))
                     return true;
-                }
             }
             return false;
         }
 
-        private unsafe void UpdateState(string directory, ref FileChangeList changes, ref WIN32_FIND_DATAW file, char* filename)
+        internal void UpdateState(string directory, ref FileChangeList changes, ref FileSystemEntry file)
         {
-            int index = _state.IndexOf(directory, filename);
+            int index = _state.IndexOf(directory, file.FileName);
             if (index == -1) // file added
             {
-                string path = new string(filename);
+                string path = file.FileName.ToString();
 
-                if (file.IsDirectory) {
-                    if (_includeSubdirectories) {
-                        _directories.Add(Path.Combine(directory.TrimEnd('*'), path, "*"));
-                    }
-                }
+                changes.AddAdded(directory, path.ToString());
 
-                changes.AddAdded(directory, path);
-                
                 var newFileState = new FileState(directory, path);
-                newFileState.LastWrite = file.LastWrite;
-                newFileState.FileSize = file.FileSize;
+                newFileState.LastWrite = file.LastWriteTimeUtc;
+                newFileState.FileSize = file.Length;
                 newFileState._version = _version;
                 _state.Add(directory, path, newFileState);
                 return;
             }
 
             _state.Values[index]._version = _version;
 
-            if (file.IsDirectory) {
-                return;
-            }
-
             var previousState = _state.Values[index];
-            if (file.LastWrite != previousState.LastWrite || file.FileSize != previousState.FileSize) {
+            if (file.LastWriteTimeUtc != previousState.LastWrite || file.Length != previousState.FileSize)
+            {
                 changes.AddChanged(directory, previousState.Path);
-                _state.Values[index].LastWrite = file.LastWrite;
-                _state.Values[index].FileSize = file.FileSize;
+                _state.Values[index].LastWrite = file.LastWriteTimeUtc;
```

Comment:
We should eventually change these to match the names that `FileSystemEntry` is using. (Which, in turn, aligns with prior System.IO terminology)